File: internal/interfaces/services/jail/jail.go
// SPDX-License-Identifier: BSD-2-Clause
//
// Copyright (c) 2025 The FreeBSD Foundation.
//
// This software was developed by Hayzam Sherif <hayzam@alchemilla.io>
// of Alchemilla Ventures Pvt. Ltd. <hello@alchemilla.io>,
// under sponsorship from the FreeBSD Foundation.

package jailServiceInterfaces

import (
	"context"

	jailModels "github.com/alchemillahq/sylve/internal/db/models/jail"
)

type HookPhase struct {
	Enabled bool   `json:"enabled"`
	Script  string `json:"script"`
}

type Hooks struct {
	Prestart  HookPhase `json:"prestart"`
	Start     HookPhase `json:"start"`
	Poststart HookPhase `json:"poststart"`
	Prestop   HookPhase `json:"prestop"`
	Stop      HookPhase `json:"stop"`
	Poststop  HookPhase `json:"poststop"`
}

type JailCreationState struct {
	CTID        uint
	DatasetName string
	JailDir     string
}

type CreateJailRequest struct {
	Name        string `json:"name" binding:"required"`
	CTID        *uint  `json:"ctId" binding:"required"`
	Hostname    string `json:"hostname"`
	Description string `json:"description"`

	Pool          string `json:"pool" binding:"required"`
	Base          string `json:"base"`
	BootstrapName string `json:"bootstrapName"`
	Fstab         string `json:"fstab"`
	ResolvConf    string `json:"resolvConf"`

	SwitchName string `json:"switchName"`

	InheritIPv4 *bool `json:"inheritIPv4"`
	InheritIPv6 *bool `json:"inheritIPv6"`

	DHCP  *bool `json:"dhcp"`
	SLAAC *bool `json:"slaac"`

	IPv4      *int   `json:"ipv4"`
	IPv4Gw    *int   `json:"ipv4Gw"`
	IPv4Raw   string `json:"ipv4Raw"`
	IPv4GwRaw string `json:"ipv4GwRaw"`

	IPv6      *int   `json:"ipv6"`
	IPv6Gw    *int   `json:"ipv6Gw"`
	IPv6Raw   string `json:"ipv6Raw"`
	IPv6GwRaw string `json:"ipv6GwRaw"`

	MAC    *int   `json:"mac"`
	MACRaw string `json:"macRaw"`

	VLAN *int `json:"vlan"`

	ResourceLimits *bool  `json:"resourceLimits"`
	Cores          *int   `json:"cores"`
	Memory         *int   `json:"memory"`
	StartAtBoot    *bool  `json:"startAtBoot"`
	StartOrder     int    `json:"startOrder"`
	DevFSRuleset   string `json:"devfsRuleset"`

	Type              jailModels.JailType `json:"type" binding:"required"`
	AllowedOptions    []string            `json:"allowedOptions"`
	CleanEnvironment  *bool               `json:"cleanEnvironment"`
	AdditionalOptions string              `json:"additionalOptions"`
	Hooks             Hooks               `json:"hooks"`

	MetadataMeta string `json:"metadataMeta"`
	MetadataEnv  string `json:"metadataEnv"`
}

type SimpleList struct {
	ID             uint   `json:"id"`
	Name           string `json:"name"`
	CTID           uint   `json:"ctId"`
	State          string `json:"state"`
	ResourceLimits *bool  `json:"resourceLimits"`
	Cores          int    `json:"cores"`
	Memory         int    `json:"memory"`
}

type SimpleTemplateList struct {
	ID             uint   `json:"id"`
	Name           string `json:"name"`
	SourceJailName string `json:"sourceJailName"`
}

type State struct {
	CTID              uint    `json:"ctId"`
	State             string  `json:"state"`
	PCPU              float64 `json:"pcpu"`
	Memory            int64   `json:"memory"`
	PendingAction     string  `json:"pendingAction,omitempty"`
	OverrideRequested bool    `json:"overrideRequested"`
}

type AddJailNetworkRequest struct {
	CTID           uint   `json:"ctId" binding:"required"`
	Name           string `json:"name" binding:"required"`
	SwitchName     string `json:"switchName" binding:"required"`
	MacID          *uint  `json:"macId"`
	MACRaw         string `json:"macRaw"`
	IP4            *uint  `json:"ip4"`
	IP4Raw         string `json:"ip4Raw"`
	IP4GW          *uint  `json:"ip4gw"`
	IP4GwRaw       string `json:"ip4gwRaw"`
	IP6            *uint  `json:"ip6"`
	IP6Raw         string `json:"ip6Raw"`
	IP6GW          *uint  `json:"ip6gw"`
	IP6GwRaw       string `json:"ip6gwRaw"`
	DHCP           *bool  `json:"dhcp"`
	SLAAC          *bool  `json:"slaac"`
	DefaultGateway *bool  `json:"defaultGateway"`
	VLAN           *int   `json:"vlan"`
}

type EditJailNetworkRequest struct {
	NetworkID      uint   `json:"networkId" binding:"required"`
	Name           string `json:"name" binding:"required"`
	SwitchName     string `json:"switchName" binding:"required"`
	MacID          *uint  `json:"macId"`
	MACRaw         string `json:"macRaw"`
	IP4            *uint  `json:"ip4"`
	IP4Raw         string `json:"ip4Raw"`
	IP4GW          *uint  `json:"ip4gw"`
	IP4GwRaw       string `json:"ip4gwRaw"`
	IP6            *uint  `json:"ip6"`
	IP6Raw         string `json:"ip6Raw"`
	IP6GW          *uint  `json:"ip6gw"`
	IP6GwRaw       string `json:"ip6gwRaw"`
	DHCP           *bool  `json:"dhcp"`
	SLAAC          *bool  `json:"slaac"`
	DefaultGateway *bool  `json:"defaultGateway"`
	VLAN           *int   `json:"vlan"`
}

type DeleteJailResult struct {
	Warnings         []string `json:"warnings"`
	RetainedDatasets []string `json:"retainedDatasets"`
}

type JailServiceInterface interface {
	JailAction(ctid int, action string) error
	ForceStopJail(ctID uint) error
	IsJailRunning(ctid uint) (bool, error)
	GetJailCTIDFromDataset(dataset string) (uint, error)
	DeleteJail(ctx context.Context, ctId uint, deleteMacs bool, deleteRootFS bool) error
	DeleteJailWithWarnings(ctx context.Context, ctId uint, deleteMacs bool, deleteRootFS bool) (DeleteJailResult, error)
	RetireJailLocalMetadata(ctx context.Context, ctId uint, deleteMacs bool) error
	StartStatsMonitoring(ctx context.Context)

	StoreJailUsage() error
	PruneOrphanedJailStats() error

	ListBootstraps(ctx context.Context, pool string) ([]BootstrapEntry, error)
	CreateBootstrap(ctx context.Context, req BootstrapRequest) error
	DeleteBootstrap(ctx context.Context, pool, name string) error
}

File: web/src/lib/api/jail/jail.ts
import {
	APIResponseSchema,
	GuestDeletionResponseSchema,
	type APIResponse,
	type GFSStep
} from '$lib/types/common';
import {
	JailActionResponseSchema,
	JailLogsSchema,
	JailRootMountPointSchema,
	JailSchema,
	JailStateSchema,
	JailStatSchema,
	JailStatsBootstrapSchema,
	JailNetworkInheritanceResultSchema,
	NetworkSchema,
	SimpleJailSchema,
	type CreateData,
	type Jail,
	type JailActionResponse,
	type JailLifecycleAction,
	type JailLogs,
	type JailRootMountPoint,
	type JailStat,
	type JailStatsBootstrap,
	type JailNetwork,
	type JailNetworkInheritanceResult,
	type JailState,
	type SimpleJail,
	JailTemplateSchema,
	type JailTemplate,
	JailTemplateCaptureTaskResponseSchema,
	type JailTemplateCaptureTaskResponse,
	JailTemplateInstantiationTaskResponseSchema,
	type JailTemplateInstantiationTaskResponse,
	SimpleJailTemplateSchema,
	type SimpleJailTemplate
} from '$lib/types/jail/jail';
import { apiRequestData, apiRequestResult, type NodeAPIDataRequestOptions } from '$lib/utils/http';
import { z } from 'zod/v4';

export async function newJail(data: CreateData, hostname?: string): Promise<APIResponse> {
	return await apiRequestResult(
		'/jail',
		APIResponseSchema,
		'POST',
		{
			name: data.name,
			hostname: data.hostname,
			ctId: Number(data.id.toString()),
			description: data.description,
			pool: data.storage.pool,
			base: data.storage.base,
			bootstrapName: data.storage.bootstrapName,
			fstab: data.storage.fstab,
			resolvConf: data.network.resolvConf,
			switchName: data.network.switch,
			dhcp: data.network.dhcp,
			slaac: data.network.slaac,
			inheritIPv4: data.network.inheritIPv4,
			inheritIPv6: data.network.inheritIPv6,
			ipv4: data.network.ipv4,
			ipv4Raw: data.network.ipv4Raw,
			ipv4Gw: data.network.ipv4Gateway,
			ipv4GwRaw: data.network.ipv4GatewayRaw,
			ipv6: data.network.ipv6,
			ipv6Raw: data.network.ipv6Raw,
			ipv6Gw: data.network.ipv6Gateway,
			ipv6GwRaw: data.network.ipv6GatewayRaw,
			mac: data.network.mac,
			macRaw: data.network.macRaw,
			vlan: Number(data.network.vlan),
			resourceLimits: data.hardware.resourceLimits,
			cores: Number(data.hardware.cpuCores.toString()),
			memory: Number(data.hardware.ram.toString()),
			startAtBoot: data.hardware.startAtBoot,
			startOrder: Number(data.hardware.bootOrder),
			devfsRuleset: data.hardware.devfsRuleset,
			jailType: data.advanced.jailType,
			additionalOptions: data.advanced.additionalOptions,
			allowedOptions: data.advanced.allowedOptions,
			hooks: data.advanced.execScripts,
			cleanEnvironment: data.advanced.cleanEnvironment,
			type: data.advanced.jailType,
			metadataMeta: data.advanced.metadata.meta,
			metadataEnv: data.advanced.metadata.env
		},
		{ hostname }
	);
}

export async function getSimpleJails(
	hostname?: string,
	signal?: AbortSignal
): Promise<SimpleJail[]> {
	return await apiRequestData('/jail/simple', z.array(SimpleJailSchema), 'GET', undefined, {
		hostname,
		signal
	});
}

export async function getSimpleJailTemplates(hostname?: string): Promise<SimpleJailTemplate[]> {
	return await apiRequestData(
		'/jail/templates',
		z.array(SimpleJailTemplateSchema),
		'GET',
		undefined,
		{ hostname }
	);
}

export async function getJailTemplateById(
	templateId: number,
	hostname?: string
): Promise<JailTemplate | APIResponse> {
	return await apiRequestResult(
		`/jail/templates/${templateId}`,
		JailTemplateSchema,
		'GET',
		undefined,
		{ hostname }
	);
}

export async function getJails(hostname?: string, signal?: AbortSignal): Promise<Jail[]> {
	return await apiRequestData('/jail', z.array(JailSchema), 'GET', undefined, {
		hostname,
		signal
	});
}

export async function getJailsResult(
	options?: NodeAPIDataRequestOptions
): Promise<Jail[] | APIResponse> {
	return await apiRequestResult('/jail', z.array(JailSchema), 'GET', undefined, options);
}

export async function getJailByCTID(
	ctId: number,
	options?: NodeAPIDataRequestOptions
): Promise<Jail | APIResponse> {
	return await apiRequestResult(`/jail/${ctId}`, JailSchema, 'GET', undefined, options);
}

export async function getJailRootMountPoint(
	ctId: number,
	options?: NodeAPIDataRequestOptions
): Promise<JailRootMountPoint | APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/root-mountpoint`,
		JailRootMountPointSchema,
		'GET',
		undefined,
		options
	);
}

export async function getSimpleJailByCTID(
	ctId: number,
	options?: NodeAPIDataRequestOptions
): Promise<SimpleJail | APIResponse> {
	return await apiRequestResult(
		`/jail/simple/${ctId}`,
		SimpleJailSchema,
		'GET',
		undefined,
		options
	);
}

export async function deleteJail(
	ctId: number,
	deleteMacs: boolean,
	deleteRootFs: boolean,
	hostname?: string
): Promise<APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}?deletemacs=${deleteMacs}&deleterootfs=${deleteRootFs}`,
		GuestDeletionResponseSchema,
		'DELETE',
		undefined,
		{ hostname }
	);
}

export async function getJailState(
	ctId: number,
	options?: NodeAPIDataRequestOptions
): Promise<JailState | APIResponse> {
	return await apiRequestResult(`/jail/${ctId}/state`, JailStateSchema, 'GET', undefined, options);
}

export async function jailAction(
	ctId: number,
	action: JailLifecycleAction,
	hostname?: string
): Promise<JailActionResponse | APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/actions/${action}`,
		JailActionResponseSchema,
		'POST',
		undefined,
		{
			hostname
		}
	);
}

export interface ConvertJailToTemplateRequest {
	name: string;
}

export async function convertJailToTemplate(
	ctId: number,
	data: ConvertJailToTemplateRequest,
	hostname?: string
): Promise<JailTemplateCaptureTaskResponse | APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/templates`,
		JailTemplateCaptureTaskResponseSchema,
		'POST',
		data,
		{ hostname }
	);
}

export interface CreateJailFromTemplateRequest {
	mode: 'single' | 'multiple';
	ctid?: number;
	name?: string;
	startCtid?: number;
	count?: number;
	namePrefix?: string;
	pool?: string;
}

export async function createJailFromTemplate(
	templateId: number,
	data: CreateJailFromTemplateRequest,
	hostname?: string
): Promise<JailTemplateInstantiationTaskResponse | APIResponse> {
	return await apiRequestResult(
		`/jail/templates/${templateId}/jails`,
		JailTemplateInstantiationTaskResponseSchema,
		'POST',
		data,
		{ hostname }
	);
}

export async function deleteJailTemplate(
	templateId: number,
	hostname?: string
): Promise<APIResponse> {
	return await apiRequestResult(
		`/jail/templates/${templateId}`,
		APIResponseSchema,
		'DELETE',
		undefined,
		{ hostname }
	);
}

export async function updateDescription(
	ctId: number,
	description: string,
	hostname?: string
): Promise<APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/description`,
		APIResponseSchema,
		'PATCH',
		{ description },
		{ hostname }
	);
}

export async function updateName(
	ctId: number,
	name: string,
	hostname?: string
): Promise<APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/name`,
		APIResponseSchema,
		'PATCH',
		{ name },
		{ hostname }
	);
}

export async function getJailLogs(
	ctId: number,
	options?: NodeAPIDataRequestOptions
): Promise<JailLogs | APIResponse> {
	return await apiRequestResult(`/jail/${ctId}/logs`, JailLogsSchema, 'GET', undefined, options);
}

export async function getStats(
	ctId: number,
	step: GFSStep,
	options?: NodeAPIDataRequestOptions
): Promise<JailStat[] | APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/stats/${step}`,
		z.array(JailStatSchema),
		'GET',
		undefined,
		options
	);
}

export async function getStatsBootstrap(
	ctId: number,
	options?: NodeAPIDataRequestOptions
): Promise<JailStatsBootstrap | APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/stats`,
		JailStatsBootstrapSchema,
		'GET',
		undefined,
		options
	);
}

export interface JailNetworkWriteRequest {
	name: string;
	switchName: string;
	macId: number;
	macRaw: string;
	ip4: number;
	ip4Raw: string;
	ip4gw: number;
	ip4gwRaw: string;
	ip6: number;
	ip6Raw: string;
	ip6gw: number;
	ip6gwRaw: string;
	dhcp: boolean;
	slaac: boolean;
	defaultGateway: boolean;
	vlan: number;
}

export async function addNetwork(
	ctId: number,
	data: JailNetworkWriteRequest,
	options?: NodeAPIDataRequestOptions
): Promise<JailNetwork | APIResponse> {
	return await apiRequestResult(`/jail/${ctId}/networks`, NetworkSchema, 'POST', data, options);
}

export async function updateNetwork(
	ctId: number,
	networkId: number,
	data: Partial<JailNetworkWriteRequest>,
	options?: NodeAPIDataRequestOptions
): Promise<JailNetwork | APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/networks/${networkId}`,
		NetworkSchema,
		'PATCH',
		data,
		options
	);
}

export async function deleteNetwork(
	ctId: number,
	networkId: number,
	options?: NodeAPIDataRequestOptions
): Promise<APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/networks/${networkId}`,
		APIResponseSchema,
		'DELETE',
		undefined,
		options
	);
}

export async function setNetworkInheritance(
	ctId: number,
	ipv4: boolean,
	ipv6: boolean,
	options?: NodeAPIDataRequestOptions
): Promise<JailNetworkInheritanceResult | APIResponse> {
	return await apiRequestResult(
		`/jail/${ctId}/network/inheritance`,
		JailNetworkInheritanceResultSchema,
		'PUT',
		{ ipv4, ipv6 },
		options
	);
}
